### PR TITLE
[FIX] crm: fix lead/opportunity meeting scheduling

### PR DESCRIPTION
**Pull request by jvnm-dev**
Reproduce the issue:
 - In the CRM app, create a lead/opportunity.
 - Schedule a meeting for this opportunity via
   the chatter and click on create.
 - Go back, the "Meeting" stat button is now at 1
 - Schedule a meeting for this opportunity via
   the chatter and click on edit then create.
 - Go back, the "Meeting" stat button is still at 1

Cause:
When we click directly on create, the opportunity ID is in the defaults
option but when we click on edit, there is no opportunity ID.

So, this commit use the context to retrieve the concerned opportunity
and set it if it is not already set.

I splitted the solution into a private method for a better readability.

OPW-2092920

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by jasonvanmalder** [comment]
@nim-odoo

File: addons/crm/models/calendar.py [hunk 1]
```diff
@@ -7,6 +7,18 @@
...
+            This method checks if the concerned model is a CRM lead.
+            The information is not always in the defaults values,
+            this is why it is necessary to check the context too.
+        """
+        res_model = defaults.get('res_model') if not ctx else ctx.get('default_res_model')
+        res_model_id = defaults.get('res_model_id') if not ctx else ctx.get('default_res_model_id')
+
+        return res_model and res_model == 'crm.lead' or \
+            res_model_id and self.env['ir.model'].sudo().browse(res_model_id).model == 'crm.lead'
+
```
Comment (simongoffin) [inline addons/crm/models/calendar.py]: Perhaps shorter and more readable if you do everything in the function:

```python
def _is_crm_lead(self, defaults, ctx=None):
        """
            This method checks if the concerned model is a CRM lead.
            The information is not always in the defaults values,
            this is why it is necessary to check the context too.
        """
        res_model = defaults.get('res_model', False) or ctx and ctx.get('default_res_model')
        res_model_id = defaults.get('res_model_id', False) or ctx and ctx.get('default_res_model_id')

        return res_model and res_model == 'crm.lead' or res_model_id and self.env['ir.model'].sudo().browse(res_model_id).model == 'crm.lead' 
```

File: addons/crm/models/calendar.py [hunk 2]
```diff
@@ -17,9 +29,13 @@ def default_get(self, fields):
         defaults = super(CalendarEvent, self).default_get(fields)
 
         # sync res_model / res_id to opportunity id (aka creating meeting from lead chatter)
-        if 'opportunity_id' not in defaults and defaults.get('res_id') and (defaults.get('res_model') or defaults.get('res_model_id')):
-            if (defaults.get('res_model') and defaults['res_model'] == 'crm.lead') or (defaults.get('res_model_id') and self.env['ir.model'].sudo().browse(defaults['res_model_id']).model == 'crm.lead'):
-                defaults['opportunity_id'] = defaults['res_id']
+        ctx = self.env.context
+        default_res_model = defaults.get('res_id') and (defaults.get('res_model') or defaults.get('res_model_id'))
```
Comment (simongoffin) [inline addons/crm/models/calendar.py]:
```python
if 'opportunity_id' not in defaults:
    if self._is_crm_lead(defaults, ctx):
        defaults['opportunity_id'] = defaults.get('res_id', False) or ctx.get('default_res_id', False)
```

**Comment by jasonvanmalder** [comment]
@simongoffin I applied the changes and tested them!

**Comment by simongoffin** [comment]
robodoo r+

**Pull request closed without merging**